src/Bicep.Core/Semantics/Namespaces/SystemNamespaceType.cs

```diff
@@ -281,7 +283,16 @@ private static IEnumerable<FunctionOverload> GetSystemOverloads(IFeatureProvider
 
             yield return new FunctionOverloadBuilder("first")
                 // TODO even with non-literal types, some type arithmetic could be performed
-                .WithReturnResultBuilder(TryDeriveLiteralReturnType("first", LanguageConstants.Any), LanguageConstants.Any)
+                .WithReturnResultBuilder((binder, fileResolver, diagnostics, arguments, argumentTypes) =>
+                {
+                    return argumentTypes[0] switch
+                    {
+                        TypedArrayType typedArrayType => new(typedArrayType.Item.Type),
+                        TupleType tupleType when tupleType.Items.Length == 0 => new(LanguageConstants.Null),
+                        _ => TryDeriveLiteralReturnType("first", LanguageConstants.Any)
+                                .Invoke(binder, fileResolver, diagnostics, arguments, argumentTypes)
```

Comment:
If you're already handling both cases for `TupleType`, the `TryDeriveLiteralReturnType` isn't providing any additional information. If `ArrayType` is unconditionally matched in an earlier arm of the expression, the `_` arm can just throw an `InvalidOperationException`, as it won't be reachable.